File: 207. Course Schedule.go
func canFinish(numCourses int, prerequisites [][]int) bool {
    inDegrees := make([]int, numCourses)
    adj := make([][]int, numCourses)
    q := []int{}
    for _, v := range prerequisites {
        inDegrees[v[0]]++
        if adj[v[1]] == nil {
            adj[v[1]] = make([]int, 0)
        }
        adj[v[1]] = append(adj[v[1]], v[0])
    }
    for i, v := range inDegrees {
        if v == 0 {
            q = append(q, i)
        }
    }
    count := 0
    for len(q) > 0 {
        current := q[0]
        count++
        q = q[1:]
        for _, next := range adj[current] {
            inDegrees[next]--
            if inDegrees[next] == 0 {
                q = append(q, next)
            }
        }
    }
    return count == numCourses
}

File: 207. Course Schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] i : prerequisites) {
            int from = i[1];
            int to = i[0];
            graph.computeIfAbsent(from, k -> new ArrayList<>()).add(to);
        }
        int[] inDegrees = new int[numCourses];
        for (int[] i : prerequisites) {
            inDegrees[i[0]]++;
        }
        Deque<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegrees[i] == 0) q.addLast(i);
        }
        int count = 0;
        while (!q.isEmpty()) {
            int current = q.pollFirst();
            count++;
            if (!graph.containsKey(current)) continue;
            for (int i : graph.get(current)) {
                inDegrees[i]--;
                if (inDegrees[i] == 0) q.addLast(i);
            }
        }
        return count == numCourses;
    }
}